--- networking/ping6.c.orig	2006-07-01 06:42:02.000000000 +0800
+++ networking/ping6.c	2011-06-27 22:28:26.000000000 +0800
@@ -310,8 +310,10 @@
 		if (options & O_QUIET)
 			return;
 
+/* purpose     : CONFIG_IPV6    author : ChunRu    date : 2011-06-27    */	
+/* description : source address should be set to "from" (for multicast) */
 		printf("%d bytes from %s: icmp6_seq=%u", sz,
-			   inet_ntop(AF_INET6, &pingaddr.sin6_addr,
+			   inet_ntop(AF_INET6, &from->sin6_addr,
 						 buf, sizeof(buf)),
 			   icmppkt->icmp6_seq);
 		printf(" ttl=%d time=%lu.%lu ms", hoplimit,
@@ -381,8 +383,10 @@
 	setsockopt(pingsock, SOL_RAW, IPV6_CHECKSUM, (char *) &sockopt,
 			   sizeof(sockopt));
 
+/* purpose     : CONFIG_IPV6    author : ChunRu    date : 2011-06-27 */	
+/* description : recv. hoplimit flag should be IPV6_RECVHOPLIMIT     */
 	sockopt = 1;
-	setsockopt(pingsock, SOL_IPV6, IPV6_HOPLIMIT, (char *) &sockopt,
+	setsockopt(pingsock, SOL_IPV6, IPV6_RECVHOPLIMIT, (char *) &sockopt,
 			   sizeof(sockopt));
 
 	if (ifname) {
@@ -390,10 +394,10 @@
 			bb_error_msg_and_die("%s: invalid interface name", ifname);
 	}
 
-	printf("PING %s (%s): %d data bytes\n",
-	           hostent->h_name,
-			   inet_ntop(AF_INET6, &pingaddr.sin6_addr,
-						 buf, sizeof(buf)),
+/* purpose     : CONFIG_IPV6    author : ChunRu    date : 2011-06-27     */	
+/* description : 1st variable set to host avoiding from domain name NULL */
+	printf("PING %s (%s): %d data bytes\n", host,
+		   inet_ntop(AF_INET6, &pingaddr.sin6_addr, buf, sizeof(buf)),
 		   datalen);
 
 	signal(SIGINT, pingstats);
